Suppress ForwardDeclChecker for headers generated by Scripts/webkit/messages.py

fred-wang opened the pull request:
#### 1e727c2cb404763c4acba240b5d98395f3312b0e
<pre>
Suppress ForwardDeclChecker for headers generated by Scripts/webkit/messages.py
<a href="https://bugs.webkit.org/show_bug.cgi?id=295621">https://bugs.webkit.org/show_bug.cgi?id=295621</a>

Reviewed by Ryosuke Niwa.

Message types are only instantiated temporarily to encode / decode bytes,
there is no inherent unsafety in using raw references or pointers there.
This patch introduces new macros to suppress ForwardDeclChecker errors
and use them for *Messages.h headers generated by the messages.py script.

* Source/WTF/wtf/Compiler.h: Introduce new macros to suppress
  alpha.webkit.ForwardDeclChecker warnings for class members and
  function arguments. This checker is available in the webkit branch of
  Ryosuke&apos;s llvm fork, was merged in upstream llvm via
  <a href="https://github.com/llvm/llvm-project/pull/130554">https://github.com/llvm/llvm-project/pull/130554</a> but is not currently
  available in the latest release of llvm (20.x) or in Apple&apos;s clang version.
* Source/WebKit/SaferCPPExpectations/ForwardDeclCheckerExpectations: Remove
  *Messages.h files.
* Source/WebKit/SaferCPPExpectations/NoUncheckedPtrMemberCheckerExpectations: Ditto.
* Source/WebKit/SaferCPPExpectations/NoUncountedMemberCheckerExpectations: Ditto.
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations: Ditto.
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations: Ditto.
* Source/WebKit/Scripts/webkit/messages.py:
(function_parameter_type): Move builtin_types array out of this function, so it
  can be used by function_parameter_requires_suppress_forward_decl.
(function_parameter_requires_suppress_forward_decl): New helper function to
  indicate parameters that require to suppress ForwardDeclChecker error.
  Essentially those for which function_parameter_type() returns a reference
  and that are not listed in type in types_that_cannot_be_forward_declared().
(message_to_struct_declaration): Add SUPPRESS_FORWARD_DECL_MEMBER in front of
  m_parameter declarations and SUPPRESS_FORWARD_DECL_ARG macros in front of
  of &apos;encoder &lt;&lt; m_parameter&apos; instructions.
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessages.h: Regenerate expectation.
* Source/WebKit/Scripts/webkit/tests/TestWithDeferSendingOptionMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithSpanOfConstMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h: Ditto.

Canonical link: <a href="https://commits.webkit.org/297178@main">https://commits.webkit.org/297178@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/585f61b8f5e160ff00a1d45ea3213ff177212868

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110714 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30373 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20809 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116741 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60981 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31052 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38963 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84185 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113662 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24805 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99699 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64626 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/110146 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24173 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17835 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60535 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103206 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94196 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17894 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119531 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109268 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37754 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28054 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93148 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38130 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95965 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92972 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23708 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38004 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15752 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33732 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37650 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43122 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/133543 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37312 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36073 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40651 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39020 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fred-wang commented on Source/WebKit/SaferCPPExpectations/NoUncheckedPtrMemberCheckerExpectations:
```diff
@@ -1,3 +1,2 @@
-EventDispatcherMessages.h
```
@rniwa @cdumez 

IIUC, this and other NoUncheckedPtrMemberCheckerExpectations/NoUncountedMemberCheckerExpectations/UncheckedCallArgsCheckerExpectations/UncountedCallArgsCheckerExpectations fixes are because at the end all what we do is setting the `[[clang::suppress]]` attribute, the macro does not really use the "alpha.webkit.ForwardDeclChecker" argument.

It seems the only remaining failure after this patch is then `alpha.webkit.NoUncheckedPtrMemberChecker` for WebPageMessages.h:

> /Volumes/Data/worker/macOS-Safer-CPP-Checks-EWS/build/WebKitBuild/Release/DerivedSources/WebKit/WebPageMessages.h:770:5: warning: Member variable 'm_event' in 'Messages::WebPage::KeyEvent' is a reference to CheckedPtr capable type 'WebKit::WebKeyboardEvent'; member variables must be a CheckedPtr, CheckedRef, WeakRef, or WeakPtr [alpha.webkit.NoUncheckedPtrMemberChecker]

but this patch does not set the `[[clang::suppress]]` attribute here because `types_that_cannot_be_forward_declared()` returns true for `'WebKit::WebKeyboardEvent`, so `function_parameter_requires_suppress_forward_decl()` returns false.  Probably we should still suppress the warning in that case.

rniwa commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -280,7 +289,10 @@ def message_to_struct_declaration(receiver, message):
-                result.append('        encoder << m_%s;\n' % message.parameters[i].name)
+                result.append('        ')
+                if requires_suppress_forward_decl[i]:
+                    result.append('SUPPRESS_FORWARD_DECL_ARG ')
```
Why do we need this here? It's definitely not a forward declaration.

fred-wang replied:
@rniwa I don't understand what you mean here? This is a function argument that is forward declared as you documented at https://clang.llvm.org/docs/analyzer/checkers.html#alpha-webkit-forwarddeclchecker

See for example https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/297168@main%20(4179)/scan-build-output/StaticAnalyzer/WebKit/StaticAnalyzerReports/report-WebPageMessages.h-encode-1-327a87.html#EndPath

rniwa commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -290,16 +302,24 @@ def message_to_struct_declaration(receiver, message):
+        parameter = message.parameters[i]
+        result.append('        ')
+        if requires_suppress_forward_decl[i]:
+            result.append('SUPPRESS_FORWARD_DECL_ARG ')
```
Again, this is not a forward declaration. What are we suppressing?

fred-wang replied:
Same here.

rniwa reviewed (approved):
I have a much simpler PR in https://github.com/WebKit/WebKit/pull/47744 but this is fine too.

fred-wang committed:
Suppress ForwardDeclChecker for headers generated by Scripts/webkit/messages.py
https://bugs.webkit.org/show_bug.cgi?id=295621

Reviewed by Ryosuke Niwa.

Message types are only instantiated temporarily to encode / decode bytes,
there is no inherent unsafety in using raw references or pointers there.
This patch introduces new macros to suppress ForwardDeclChecker errors
and use them for *Messages.h headers generated by the messages.py script.

* Source/WTF/wtf/Compiler.h: Introduce new macros to suppress
  alpha.webkit.ForwardDeclChecker warnings for class members and
  function arguments. This checker is available in the webkit branch of
  Ryosuke's llvm fork, was merged in upstream llvm via
  https://github.com/llvm/llvm-project/pull/130554 but is not currently
  available in the latest release of llvm (20.x) or in Apple's clang version.
* Source/WebKit/SaferCPPExpectations/ForwardDeclCheckerExpectations: Remove
  *Messages.h files.
* Source/WebKit/SaferCPPExpectations/NoUncheckedPtrMemberCheckerExpectations: Ditto.
* Source/WebKit/SaferCPPExpectations/NoUncountedMemberCheckerExpectations: Ditto.
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations: Ditto.
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations: Ditto.
* Source/WebKit/Scripts/webkit/messages.py:
(function_parameter_type): Move builtin_types array out of this function, so it
  can be used by function_parameter_requires_suppress_forward_decl.
(function_parameter_requires_suppress_forward_decl): New helper function to
  indicate parameters that require to suppress ForwardDeclChecker error.
  Essentially those for which function_parameter_type() returns a reference
  and that are not listed in type in types_that_cannot_be_forward_declared().
(message_to_struct_declaration): Add SUPPRESS_FORWARD_DECL_MEMBER in front of
  m_parameter declarations and SUPPRESS_FORWARD_DECL_ARG macros in front of
  of 'encoder << m_parameter' instructions.
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessages.h: Regenerate expectation.
* Source/WebKit/Scripts/webkit/tests/TestWithDeferSendingOptionMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithSpanOfConstMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessages.h: Ditto.
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h: Ditto.

Canonical link: https://commits.webkit.org/297178@main
